python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -162,6 +166,22 @@ def __init__(
         if isinstance(node_def, GraphDefinition):
             _validate_graph_def(node_def)
 
+        if node_def is None:
+            check.invariant(
+                len(keys_by_input_name) == 0,
+                "node_def is None, so keys_by_input_name must be empty",
+            )
+            check.invariant(
+                len(keys_by_output_name) == 0,
+                "node_def is None, so keys_by_output_name must be empty",
+            )
+            check.invariant(
+                backfill_policy is None, "node_def is None, so backfill_policy must be None"
+            )
+            check.invariant(
+                not can_subset, "node_def is None, so backfill_policy must not be provided"
+            )
+
```

Comment:
It would be very nice to wrap up all these into a single all or something object internally. This demonstrates a violation of [Invalid parameter combinations](https://github.com/dagster-io/internal/discussions/9694). While we can't break callers we can restructure the internals. We can do this as a follow on this is a pre-existing problem. It's just a good demo of that this change is excellent because is more obviously exposed the _existing_ smell.